src/renderer/src/hooks/useMapeoDeviceList.ts

```diff
@@ -0,0 +1,27 @@
+import { useMapeoDeviceStore } from './stores/mapeoDeviceStore'
+
+export const useMapeoDeviceMembersListIds = () => {
+  const devices = useMapeoDeviceStore(
+    (store) => store.memberDevices,
+    (oldVal, newVal) => {
+      const oldValArray = Object.keys(oldVal).map((key) => key)
+      const newValArray = Object.keys(newVal).map((key) => key)
+      return oldValArray !== newValArray
```

Comment:
this will always return true due to array equality by reference instead of value. what you want is a deep equality check like https://lodash.com/docs/4.17.15#isEqual

Reply:
do you need to provide your own equality checker to begin with?

Reply:
In this case we actually do! 

The reason is if we change the role of one of the members, that member will still still in the same list: `memberDevice`. since were returning the entire list in this case, the previous rendered list !== new List, causing a re-render, and therefore anything using the list will re-render. 

This is a general problem with react in this case, not zustand. 

So to get around this, we can only check the equallity of the list of ids, instead of the list of objects.

Reply:
but isnt that the desired behavior if you're returning the whole list with this selector? anything that subscribes to the memberDevices list should re-render whenever any member in that list changes 🤔 (similar for plain react state)

Reply:
No, agreed. But in our case, the list is always open due to the use of our modals. So any edit made to any member will cause alot of extra rerenders.

So my solution to this, is to only return a list of `ids` to iterate through the list. And then use the `useMemberDevice(id)` hook, which is memoized to the specific device. this method will allow us to have a list of devices, where each object on the list is not re-rendered on every change to another member in the list

Reply:
hm I don't think that optimization should happen here in that case. probably one of the following would seem more appropriate:

- do the memoization in the consuming component i.e. the modals, using a useMemo.

- use a different selector that only returns an array of ids, if that's the only thing of interest (seems unlikely)

from this selector, I'd expect a mapping of device IDs to their respective and _most recent_ device state. your approach doesnt account for the last case because you wouldn't update the returned list object of a device has a property updated e.g. its role.

Reply:
this function does only return ids!

Memoizing the consuming component is a little difficult in this use case is a little difficult due to object equality and zustand being an immutable store. in JS. the consuming component would need to memoize based off each property would get labrious

Reply:
think we're talking about memoizing different things. you're saying that you only want the modals to re-render when a device is added or removed from the list. if that's the case, the memoization in the modal could look something like:

```ts
const members = useMapeoDevices()

// the component would only use this in the render
const memoizedMembers = useMemo(() => {
  // here you would use that equality check on the array of ids 
  // to determine if this returns a new value
}, [members])
```

Reply:
yeah! I tried that originally. I don't fully understand the inner working of the store, but for some reason its not memoized automatically, we still have to do the equality check.  We could do this though

```ts
export const useMapeoDeviceMembersListIds = () => {
  const devices = useMapeoDeviceStore(
    (store) => Object.keys(store.memberDevices).map((key) => key),
    (oldVal, newVal) =>  JSON.stringify(oldVal) !== JSON.stringify(newVal)
  )

  return Object.keys(devices).map(([key]) => key)
}

```

Reply:
hm interesting, but yeah what you provide seems fine. wasn't clear when I initially commented on this that this hook should only returning an array of ids, hence the deep dive 😄